### ISSUE TITLE
Refactor: Introduce DataConfig and restructure ModelConfig hierarchy

### ISSUE THREAD

**Issue by ShivamSingal** [Blood-Glucose-Control/nocturnal-hypo-gly-prob-forecast#397]
## Summary

Our `ModelConfig` base class has grown to 27 fields, many of which are unused dead weight or domain-specific data fields duplicated across model configs. A deep research survey of 6 major time series forecasting libraries reveals a consistent pattern: **separate data/domain configuration from model architecture configuration**. This issue proposes introducing a `DataConfig` class and slimming down `ModelConfig` to follow industry best practices.

This restructuring also unblocks a concrete feature need: adding **known future covariates** (e.g., `hour_sin`, `hour_cos` time-of-day features) to model configs. Currently there's no clean place to put `known_covariate_cols` because even `covariate_cols` (past-only) isn't in the base — it's duplicated identically in Chronos2Config and TiDEConfig while TTM uses a different name (`input_features`) for the same concept.

---

### Base `ModelConfig` (27 fields)

**File:** `src/models/base/base_model.py`

The base config mixes three concerns:
1. **Universal model identity** (7 fields actually used by all models)
2. **HuggingFace Trainer scheduling** (6 fields used by zero subclasses)
3. **Generic training params** (14 fields used by 1-2 models each)

#### Fields in base that NO subclass uses (dead weight)

| Field | Category | Verdict |
|---|---|---|
| `d_model` (512) | Architecture | No subclass references it — each model has its own arch params |
| `n_heads` (8) | Architecture | Same — dead weight |
| `n_layers` (6) | Architecture | Same — dead weight |
| `eval_strategy` ("steps") | HF Trainer | No subclass uses directly |
| `eval_steps` (1000) | HF Trainer | No subclass uses directly |
| `save_steps` (2000) | HF Trainer | No subclass uses directly |
| `logging_steps` (100) | HF Trainer | No subclass uses directly |
| `metric_for_best_model` ("eval_loss") | HF Trainer | No subclass uses directly |
| `greater_is_better` (False) | HF Trainer | No subclass uses directly |
| `fp16` (True) | Backend-specific | TiDE uses `precision: "16-mixed"`, TimesFM uses `torch_dtype: "bfloat16"` — each backend has its own precision mechanism |

### Data/Domain Fields Duplicated Across Model Configs

These fields appear identically in multiple model configs but are NOT in the base:

| Field | Default | Defined In | TTM Equivalent |
|---|---|---|---|
| `covariate_cols` | `["iob"]` | Chronos2Config, TiDEConfig | `input_features` (different name, different type) |
| `target_col` | `"bg_mM"` | Chronos2Config, TiDEConfig | `target_features` (List[str] vs str) |
| `patient_col` | `"p_num"` | Chronos2Config, TiDEConfig | hardcoded in column specifiers |
| `time_col` | `"datetime"` | Chronos2Config, TiDEConfig | hardcoded in column specifiers |
| `interval_mins` | `5` | Chronos2Config, TiDEConfig | `resolution_min` (different name) |
| `imputation_threshold_mins` | `45` | Chronos2Config, TiDEConfig, TimesFMConfig | not in TTM |
| `min_segment_length` | `None` (auto) | Chronos2Config, TiDEConfig | not in TTM |

### Same Concept, Different Names

| Concept | Names Used | Where |
|---|---|---|
| Learning rate | `learning_rate`, `fine_tune_lr`, `lr` | Base/TTM/TimesFM, Chronos2, TiDE |
| Sampling interval | `interval_mins`, `resolution_min`, `freq` | Chronos2/TiDE, TTM, TimeGrad |
| Target column | `target_col` (str), `target_features` (List[str]) | Chronos2/TiDE, TTM |
| Covariates | `covariate_cols`, `input_features` | Chronos2/TiDE, TTM |
| Batch size | `batch_size`, `fine_tune_batch_size` | Base/TTM/TiDE, Chronos2 |
| Model checkpoint | `model_path`, `checkpoint_path` | Base, TimesFM (alias) |
| Forecast horizon | `forecast_length`, `horizon_length` | Base, TimesFM (alias) |

---

### Survey of 6 Major Time Series Libraries

Every surveyed library separates data configuration from model configuration:

#### 1. AutoGluon TimeSeries
- **Data config** lives on `TimeSeriesPredictor`: `prediction_length`, `target`, `known_covariates_names`, `eval_metric`, `freq`
- **Model config** is untyped dicts passed to `fit(hyperparameters={"Chronos2": {...}})`
- **Covariates**: `known_covariates_names` (future-known) set at Predictor level, not per-model
- [Docs: TimeSeriesPredictor API](https://auto.gluon.ai/stable/api/autogluon.timeseries.TimeSeriesPredictor.html)
- [Docs: Forecasting In Depth](https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-indepth.html)

#### 2. NeuralForecast (Nixtla)
- **Base class** (`BaseWindows`) has shared params: `h`, `input_size`, `learning_rate`, `batch_size`, `max_steps`, `scaler_type`
- **Covariates**: Three explicit lists — `hist_exog_list` (past-only), `futr_exog_list` (known future), `stat_exog_list` (static)
- **Architecture params** live on each model subclass
- [Docs: How to Add Models](https://nixtlaverse.nixtla.io/neuralforecast/examples/how_to_add_models.html)
- [Source: NeuralForecast GitHub](https://github.com/Nixtla/neuralforecast)

#### 3. IBM TSFM (TTM's upstream)
- **Model config**: `TinyTimeMixerConfig` extends HF `PretrainedConfig` — architecture only (`d_model`, `num_layers`, `patch_length`)
- **Data config**: `TimeSeriesPreprocessor` — separate class with `target_columns`, `observable_columns`, `control_columns`, `conditional_columns`, `static_categorical_columns`
- **Cleanest separation** of all surveyed libraries — preprocessor travels with model at save/load time
- [Source: TTM Tutorial](https://github.com/IBM/tsfm/blob/main/notebooks/tutorial/ttm_tutorial.ipynb)
- [HuggingFace: granite-timeseries-ttm-r2](https://huggingface.co/ibm-granite/granite-timeseries-ttm-r2)

#### 4. Darts (unit8)
- **Covariate type encoded in class hierarchy**: `PastCovariatesTorchModel`, `FutureCovariatesTorchModel`, `DualCovariatesTorchModel`, `MixedCovariatesTorchModel`
- **Shared base** (`TorchForecastingModel`): `input_chunk_length`, `output_chunk_length`, `batch_size`, `n_epochs`
- **Data** passed at `fit()`/`predict()` time, not configured upfront
- [Docs: Torch Forecasting Models](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html)
- [Source: TorchForecastingModel](https://github.com/unit8co/darts/blob/master/darts/models/forecasting/torch_forecasting_model.py)

#### 5. GluonTS
- **Minimal base**: `Estimator` has only `prediction_length`, `freq`, `lead_time`
- **PyTorch base** adds `trainer_kwargs`, `batch_size`, `num_batches_per_epoch`, `num_workers`
- **Covariates** defined in data transformations (`FieldName.FEAT_DYNAMIC_REAL`, etc.), not config
- [Docs: Estimator](https://ts.gluon.ai/dev/api/gluonts/gluonts.model.estimator.html)
- [Source: DeepAR Estimator](https://github.com/awslabs/gluonts/blob/dev/src/gluonts/torch/model/deepar/estimator.py)

#### 6. PyTorch Forecasting
- **Strongest separation**: `TimeSeriesDataSet` (data config with 6 covariate categories) + `BaseModel` (model config)
- **Bridge method**: `Model.from_dataset()` reads dataset metadata and auto-configures the model
- **Covariates**: `time_varying_known_reals`, `time_varying_unknown_reals`, `time_varying_known_categoricals`, `time_varying_unknown_categoricals`, `static_reals`, `static_categoricals`
- [Docs: Data](https://pytorch-forecasting.readthedocs.io/en/stable/data.html)
- [Docs: BaseModelWithCovariates](https://pytorch-forecasting.readthedocs.io/en/v1.4.0/api/pytorch_forecasting.models.base._base_model.BaseModelWithCovariates.html)

### Key Pattern Across All Libraries

| Library | Base Model Fields | Data/Covariate Config | Covariate Categories |
|---|---|---|---|
| AutoGluon | ~3 (predictor level) | Predictor-level lists | known, static |
| NeuralForecast | ~6 | Model constructor lists | hist_exog, futr_exog, stat_exog |
| TSFM | Architecture only | TimeSeriesPreprocessor | observable, control, conditional, static |
| Darts | ~4 | fit()/predict() args | past, future (class hierarchy) |
| PyTorch Forecasting | ~2 | TimeSeriesDataSet | 6 categories (known/unknown × real/cat + static) |
| GluonTS | ~3 | Data transformations | FieldName enums |

**Consensus**: Base model config should have **≤10 truly universal fields**. Data/domain/covariate config is always separate.

---

### Introduce `DataConfig` (composed into ModelConfig)

```python
@dataclass
class DataConfig:
    """Data and domain configuration — shared across all models.
    
    Defines column mappings, covariate specification, data resolution,
    and gap handling parameters. Composed into ModelConfig so all models
    share a single source of truth for data schema.
    """
    # Column mapping
    target_col: str = "bg_mM"
    patient_col: str = "p_num"
    time_col: str = "datetime"
    
    # Covariates (following NeuralForecast/TSFM naming conventions)
    past_covariate_cols: List[str] = field(default_factory=list)
    """Past-only covariates — available in context window but NOT provided
    for forecast horizon. E.g., IOB, COB (reactive to future BG on AID)."""
    
    known_covariate_cols: List[str] = field(default_factory=list)
    """Known future covariates — deterministic features available for
    the entire forecast horizon. E.g., hour_sin, hour_cos, day_of_week."""
    
    # Data grid
    interval_mins: int = 5
    
    # Gap handling
    imputation_threshold_mins: int = 45
    min_segment_length: Optional[int] = None  # auto = context + forecast
```

### Slim down `ModelConfig`

```python
@dataclass
class ModelConfig:
    """Base configuration for all time series models.
    
    Contains only truly universal fields. Model-specific architecture
    and backend parameters live in subclass configs.
    """
    # Model identity
    model_type: str = "base"
    model_path: Optional[str] = None
    
    # Task definition
    context_length: int = 512
    forecast_length: int = 96
    training_mode: str = "fine_tune"
    training_backend: TrainingBackend = TrainingBackend.TRANSFORMERS
    
    # Data configuration (composed, not inherited)
    data: DataConfig = field(default_factory=DataConfig)
    
    # Probabilistic forecasting
    quantile_levels: Optional[List[float]] = None
    
    # Common training params (used by most but not all models)
    learning_rate: float = 1e-4
    batch_size: int = 64
    num_epochs: int = 10
    weight_decay: float = 0.01
    early_stopping_patience: int = 10
    use_cpu: bool = False
    loss_function: str = "mse"
```

**Removed from base** (10 fields → model-specific or deleted):
- `d_model`, `n_heads`, `n_layers` — architecture-specific, no subclass uses them
- `eval_strategy`, `eval_steps`, `save_steps`, `logging_steps`, `metric_for_best_model`, `greater_is_better` — HF Trainer-specific, no subclass uses them
- `fp16` — each backend has its own precision mechanism
- `dataloader_num_workers` — training-backend-specific
- `warmup_steps`, `gradient_clip_val`, `freeze_backbone`, `dropout` — used by ≤2 models, move to those configs

### Model configs become cleaner

```python
@dataclass
class Chronos2Config(ModelConfig):
    """Chronos-2 config — AutoGluon backend-specific params only."""
    model_type: str = "chronos2"
    model_path: str = "autogluon/chronos-2"
    forecast_length: int = 72
    training_backend: TrainingBackend = TrainingBackend.CUSTOM
    
    # AutoGluon-specific (no data fields — they're in self.data)
    fine_tune_steps: int = 15000
    fine_tune_lr: float = 1e-5
    fine_tune_batch_size: int = 32
    eval_metric: str = "WQL"
    enable_ensemble: bool = False
    eval_during_fine_tune: bool = True
    min_past: int = 1
    time_limit: Optional[int] = None
```

No more `target_col`, `patient_col`, `time_col`, `covariate_cols`, `interval_mins`, `imputation_threshold_mins` — all read from `self.data.*`.

---

### Phase 1: Introduce `DataConfig` alongside existing fields (backwards-compatible)

1. Create `DataConfig` dataclass in `src/models/base/data_config.py`
2. Add `data: DataConfig` field to `ModelConfig` with a default
3. Add a `__post_init__` bridge that populates `DataConfig` from legacy per-model fields (e.g., if `self.covariate_cols` exists, copy to `self.data.past_covariate_cols`)
4. Update all model code to read from `self.config.data.*` instead of `self.config.target_col`, etc.

### Phase 2: Remove duplicated fields from model configs

1. Remove `target_col`, `patient_col`, `time_col`, `covariate_cols`, `interval_mins`, `imputation_threshold_mins`, `min_segment_length` from Chronos2Config and TiDEConfig
2. Remove `input_features`, `target_features`, `resolution_min` from TTMConfig
3. Update YAML configs to nest data fields under `data:` key
4. Update all scripts that construct configs

### Phase 3: Slim down base ModelConfig

1. Remove unused fields (`d_model`, `n_heads`, `n_layers`, HF Trainer scheduling fields)
2. Move `warmup_steps`, `gradient_clip_val`, `dropout`, `freeze_backbone` to the 1-2 model configs that use them

### Phase 4: Wire known future covariates

1. Each model reads `self.config.data.known_covariate_cols`
2. AutoGluon models (Chronos-2, TiDE): pass as `known_covariates_names` to `TimeSeriesPredictor`, generate future values at predict time
3. TTM: move known covariates from `observable_columns` to `control_columns`, pass `future_time_series` at predict time
4. Shared utility `generate_future_known_covariates()` in `src/data/preprocessing/feature_engineering.py` computes deterministic future values (sin/cos from timestamps)
5. Models that don't support covariates (TimesFM): log warning and ignore

---

### 1. Composition over inheritance for DataConfig

`ModelConfig` **contains** a `DataConfig` rather than inheriting from it. This follows TSFM's pattern (`TimeSeriesPreprocessor` is a separate object) and avoids:
- Diamond inheritance if we later add a `TrainingConfig` mixin
- Confusion about which class "owns" data fields
- The antipattern of a 40+ field god-class

### 2. Covariate taxonomy follows NeuralForecast convention

Three categories, matching the most widely adopted naming:
- **`past_covariate_cols`** (NeuralForecast: `hist_exog_list`, TSFM: `observable_columns`, Darts: `past_covariates`)
- **`known_covariate_cols`** (NeuralForecast: `futr_exog_list`, TSFM: `control_columns`, AutoGluon: `known_covariates_names`)
- **`static_cols`** (future slot — NeuralForecast: `stat_exog_list`, PyTorch Forecasting: `static_reals`)

We use `known_covariate_cols` over `future_covariate_cols` because it matches AutoGluon (our primary backend) and emphasizes that these values are **known** (deterministic), not merely "future."

### 3. Base ModelConfig shrinks to ≤15 fields

The current 27-field base violates YAGNI. Fields used by zero subclasses (`d_model`, `n_heads`, HF Trainer scheduling) are dead weight that:
- Confuse new contributors about what's "base" vs "optional"
- Create false expectations that subclasses should use them
- Make the base class harder to maintain

The slimmed base keeps only fields that ≥3 models actually reference.

### 4. Naming normalization uses the most common variant

| Concept | Chosen Name | Rationale |
|---|---|---|
| Sampling interval | `interval_mins` | Used by 2/3 models already; `resolution_min` is TTM-specific jargon |
| Past covariates | `past_covariate_cols` | Explicit temporal qualifier; avoids bare `covariate_cols` ambiguity |
| Target column | `target_col` | Simpler than TTM's `target_features: List[str]`; single-target is our use case |
| Learning rate | `learning_rate` | Python/ML convention; `lr` is an abbreviation |

### 5. Backwards compatibility via bridge `__post_init__`

Phase 1 keeps the old per-model fields working by copying them into `DataConfig`. This allows incremental migration:
- Existing YAML configs keep working
- Existing scripts keep working
- Models can be migrated one at a time
- The bridge is removed in Phase 3 once all consumers are updated

---

## Why This Is Better

1. **Single source of truth for data schema.** Currently, adding a new covariate type requires editing Chronos2Config, TiDEConfig, AND TTMConfig (with different field names). With `DataConfig`, it's one place.

2. **New models are easier to add.** A contributor adding a new model inherits `DataConfig` automatically — no need to rediscover that `target_col="bg_mM"` and `patient_col="p_num"` from copying another config.

3. **Known future covariates have a natural home.** `DataConfig.known_covariate_cols` sits alongside `past_covariate_cols` — the taxonomy is self-documenting.

4. **Base class accurately represents what's shared.** The current 27-field base creates noise. A 15-field base with only universally-used fields is honest about the contract.

5. **Consistent with industry practice.** Every major TS library separates data config from model config. We're removing a known antipattern.

---

## Related

- Time-of-day covariates (`hour_sin`, `hour_cos`) already generated in `src/data/preprocessing/feature_engineering.py` and verified on watgpu (Brown 2019 + Aleppo 2017)
- Chronos-2 time covariate experiment showed -0.6% RMSE (Brown) and -3.1% RMSE (Aleppo) improvement
- Known covariate wiring for Chronos-2/TiDE/TTM is blocked on this config restructure
